### ISSUE TITLE
Question about controlling position at which to insert element.

### ISSUE THREAD

**Issue by ih2502mk** [Matt-Esch/virtual-dom#104]
Hi,

I'm using mercury and I have an app that looks a lot like todo list i.e. a form to add new element and a list of added elements. However in my case I want to have a transition when new element is added and I want a new element to be added to the top of the list "layoutwise".
I could figure out how to use hooks to achieve transition effect - that worked perfect for me.

The problem is with the way DOM is patched when new item is added to an array of items in the app state. Transition always happens on the last item in the list.

To give you an idea, I have:
```javascript
state.items = mercury.array([]);
```
and when I "submit" a form: 
```javascript
state.items.unshift(Item( /* data that came from the form*/ ));
```
and rendering looks like this:
```javascript
function renderItemsList(state) {  
  return h('ul#item-list', [
    state.items.map(function(item) {return Item.render(item);})
  ]);
}
``` 

My assumption is that a process of creating a patch looks something like this:
```
original Vtree:
ul
 - item3
 - item2
 - item1

updated Vtree
ul
 - item4
 - item3
 - item2
 - item1

and the patch looks like this:
 - change item3 to item4
 - change item2 to item3
 - change item1 to item2
 - insert item1
```

Whereas I would like to have a patch that looks like: 
```
 - insert item4 before item3
```

I'd like to know if my assumption of how patch is created is correct to begin with. And also if there is a way to achieve the desired outcome.

Here is a requirebin of an application that I was talking about: http://requirebin.com/?gist=086fb96724577864a3b6

**Comment by AlexGalays** [comment]
Not sure, I don't know the codebase yet. It looks like you already put in a few optimizations? How does that reverse index work?

I don't really get why a new node inserted at index 3, is actually appended (INSERT patch) then moved. I guess that made other uses cases simpler?

Would a naive optimization approach work?
```
if (num(INSERT patches) == n-1) { // n = total number of patches; bail fast in most occasion
    check the MOVE patch; for each of these (new) n-1 indices, if they're being moved to a consecutive 0 index, delete all the other scheduled moves.
}
```
use-case-specific optimisations like this one may feel weird but we have to remember this is being used in UIs where prepend-the-newest-items is a very common operation (e.g Facebook top right activity panel)

For info, knockout computes the levenshtein distance between the 2 arrays to compute the least amount of patches, I don't think you can beat that at the DOM level but it's unclear whether the JS overhead makes it worth it.

**Comment by AlexGalays** [comment]
Another issue with this excessive shifting I just found; this technique cannot be implemented either:  http://blog.stevensanderson.com/2013/03/15/animating-lists-with-css-3-transitions/

Because instead of naturally flowing to the right, each item is removed/re-added, it's transform transition thus not playing.

So it plays very badly with both CSS transitions and animations.

**Comment by Raynos** [comment]
cc @Matt-Esch

**Comment by Matt-Esch** [comment]
Relying on the existence or order of DOM mutation is always wrong. It's not the right way to think. It assumes knowledge of previous state and expectation on the internals of virtual-dom, both of which we should not have to think about.

**Comment by AlexGalays** [comment]
@Matt-Esch: In an ideal world, yes; the virtual dom shouldn't be a leaky abstraction and you would just let it do whatever it needs to do without caring about how it does it.
But we cannot ignore the rest of the ecosystem either, and CSS transitions/animations get completely broken when there are phantom element moves. Js animations might be a workaround, but that would be a real shame.

**Comment by Matt-Esch** [comment]
You should be activating the CSS transitions using hooks and nextTick. That is, you should know exactly which nodes you wish to animate based on your data, and you should use that data to add a transition hook based on next tick.

This is where the flaw in this topic is; the lack of encoding of the animation state. You don't have to do animations with JS, I don't recommed it over CSS transitions, but you do need to model your expectations properly in your data model and apply transitions to the nodes using that data. Generic transitions that rely on the way in which the DOM is mutated isn't going to work consistently.

For example, if you want an inserted transition, you might add a `wasInserted` boolean flag to your model.

On rendering that item, if `wasInserted` is true, you add an animation hook which, on next tick adds a css class like `.inserted`. You code your CSS transitions against this class.

On next tick your hook will add the class and the transition will happen. Further to that you will want to clear the wasInserted flag, probably also on next tick.

There are tons of these flag situations which should not trigger re-render. I think mercury needs to add something like this as a primitive type but that's outside the scope of virtual-dom.
But as you can see, the animation works because you have recorded the state and necessity for the transition, and did not rely simply on the insertion of the node to trigger that animation.

**Comment by Raynos** [comment]
@Matt-Esch I just created #157.

We should document best practices around virtual dom and css animations in the virtual-dom repository. 

I literally just copied your comment, thats probably not quite right, but with a bit of cleanup it would make a great first piece of documentation.

**Comment by AlexGalays** [comment]
@Matt-Esch: I see where you're coming from, explicit > implicit.
Still, the cherry on top is to have css transitions and animations to implicitly work, especially for a PoC/MVP. My first approach was actually to use nextTick/equivalent, but Firefox didn't like it and I had to use a setTimeout(>20) for it to work, quite ridiculous; hence me trying other techniques.

@neonstalwart's PR is nice to have in any case.

**Comment by stevenvachon** [comment]
@AlexGalays https://github.com/YuzuJS/setImmediate

**Comment by AlexGalays** [comment]
@stevenvachon I don't remember the exact setup I had but when it works with setTimeout(..,20) but doesn't with setTimeout(.., 0),  setImmediate won't solve anything as it's faster or as fast as setTimeout(..., 0). What's more, setImmediate is a half assed construct right now with weird shims and browsers that may end up not implementing it at all or clamping it anyway.

**Comment by stevenvachon** [comment]
@AlexGalays Node.js implemented `setImmediate` and any such "weird shims" for the browser still work under the same simple API. Even if it never becomes a standard, it is nice to have access to. But I understand now how it doesn't apply to your previous comment.

**Comment by ih2502mk** [comment]
Regarding insert animation specifically you can avoid having extra flag boolean values in state that is passed to a component render function (some developers may feel uncomfortable mixing that presentational concern in). This is an example of insert animation that worked for me:
```javascript
Item.render = function(state) {
  return h('li', {
    'class' : new ItemInsertHook('item'),
  }, [
    h('div.text', state.text),    
    h('button', {
      'ev-click' : mercury.event(...)
    }, 'Remove or something...'),
  ]);
}

function ItemInsertHook(value) {
  this.value = value;
}

ItemInsertHook.prototype.hook = function(elem, propName) {
  if (!elem.childNodes.length) {
    elem.setAttribute(propName, this.value + ' inserting');

    nextTick(function () {
      elem.setAttribute(propName, this.value + '');
    }.bind(this))
  }
}

//Elswhere at the top level of application:
function renderItemsList(state) {  
  return h('ul#item-list', [
    state.items.map(function(item) {return Item.render(item);})
  ]);
}
```
And the css would look like this:
```css
li.item.inserting { opacity : 0.01; }
li.item { transition: opacity 0.2s ease-in-out; }
li.item { opacity : 0.99; }
```

When new item is added to application state virtual dom will be updated. On the next animation frame and actual item will be inserted in actual dom with value of `class` attribute being `"item inserted"`. On next tick `class` will change to just `"item"` which will invoke a transition.

I can't say if it it's possible to avoid having flag values in state but simple cases like insert transitions can totally be implemented.

Also insert hook looks like a good place to isolate dom element manipulation if you really want to keep it separate.

**Comment by Raynos** [comment]
@ih2502mk 

This is a really nice example of an animation hook. I'd strongly encourage you to contribute this idea to a `docs/insert-animations.md`

**Comment by neonstalwart** [comment]
insertion is the easiest case.  removing is more difficult - i think it requires a widget as the container for the list so that when the data indicates a node has been removed you can control the removal of the node to happen only after the animation has finished.  a widget like [`ReactCSSTransitionGroup`](http://facebook.github.io/react/docs/animation.html) would be a nice challenge to implement.

**Comment by ih2502mk** [comment]
@neonstalwart true, removing items is really a more difficult problem

I was experimenting with idea on how to remove an item through transition.
Let's say there is an event that initiates item removal like a button click. Instead of boldly removing an entry from a state, a handler for that event should initiate a transition and subscribe state change to `"transitionend"` event of that transition.

I'm experiencing some odd issues probably related to reuse of dom elements or something. Also the code gets really messy. In context of my previous comment:

```javascript
function removeItem(item) {
  var index_to_remove;
  state.items.map(function(a_item, i) {
  	if(a_item.desc() === item.desc) index_to_remove = i;
  })

  state.items.splice(index_to_remove, 1);
}


Item.render = function(state) {
  return h('li', {
    'class' : new ItemInsertHook('item'),
  }, [
    h('button.remove', {
      'ev-click' : function(ev) {
        var item_dom = ev.target.parentNode;
        item_dom.addEventListener('transitionend', function() {
          removeItem(state);
        }, false);
        item_dom.style.opacity = 0.01;
      },
    }, 'Remove'),
    h('div.desc', state.desc),
  ]);
}
```

This is a Requirebin with full example http://requirebin.com/?gist=8b20457809c1a8860d43 
There is some logic/timing flaw in order of transitions and manipulations.

**Comment by neonstalwart** [comment]
interesting... i hadn't thought about delaying updating the model until the transition ends.  i was always thinking in terms of removing the item from the model first and then using a transition in the view to reflect the change in the model but it does seem a lot simpler to do the transition in the view first and then remove the item from the model.  this seems somewhat the reverse of what normally happens - the view normally follows the model rather than lead it - but maybe it's ok.  i'll think about some more now that i have a different perspective to consider it from.

**Comment by ih2502mk** [comment]
@neonstalwart, 
that example is not the best I it get's very complicated and I don't like how `ev-transitionend` handler is strongly tied with `ev-click`. That kinda works here but in case of a more complicated component (say there would be a button group which is a whole different thing living in different module) this approach probably will not work.

Also I tried another more "model-driven" approach here http://requirebin.com/?gist=559d5f300d1a8d83523c
But again there is a lot of `setTimeout`s with almost arbitrary times and `nextTick`s.

**Comment by stevenvachon** [comment]
How would sorting (an array) be handled? Are they repositioned in the DOM, or removed+added?

**Comment by stevenvachon** [comment]
Also, what about a boolean add/remove on the same element? If the element were removed, then re-added before its transition completed, would two identical elements be visible, or would the "removing" element he "re-added" and transitions back into place?

**Comment by ih2502mk** [comment]
@stevenvachon, 
from what I've learned experimenting, sorting items in the model will not influence position of dom elements. The actual `<li />` stays in place but its contents i.e. leaf textNodes and attributes will change (or to be more precise changes will occur according to render function).

As for adding/removing: the dom element shows up on next animation frame as transparent, one frame further a class of element changes to such that it is not transparent but there is a css transition which makes an element slowly change transparency (but that is not virtual-dom's concern already). If an item is removed from the model, on next animation frame an element will disappear from the dom it can happen before transition ends. 
If there is a trick to slowly remove an item, what will happen is instead of removing an item right away remove button sets dom element's class to such that it becomes transparent, but since there is a transition it will become transparent over time and item disappear after that. So `transparent -> opaque` transition will be interrupted in the middle and `opaque (or rather semi opaque) -> transparent` transition will start instead.

**Comment by neonstalwart** [comment]
sorting depends on if you use keys.  if you use a `key` then those exact DOM nodes will be re-ordered.  if you don't use `key` then your nodes will be re-used in-place rather than swapped.

**Comment by arianvp** [comment]
I made a pull-request for Elm's virtual-dom library where we just choose a standard classname for this stuff. Maybe we should do that here as well upstream? https://github.com/evancz/virtual-dom/pull/6

**Comment by neonstalwart** [comment]
I thought more about removal and I don't think it's right for the view to update the model on transitionend. I think the model should reflect the removal immediately and something like a widget should be used to build a container that understands what to do when an item is removed from the model - i.e. keep rendering the node until transitionend.

**Comment by ih2502mk** [comment]
I tried another approach. The idea is to keep the role of `virtual-dom` to just creating and removing elements. So all transitioning happens at the level of data-structure that is being passed to render function. 
For example removing happens like this: an item that is supposed to be removed is replaced in the model with a special "fake" item that render function knows to render differently (whether it is with a class or with animation or something else). After a timeout or based on some logic the "fake" item is removed from data-structure. In this case `virtual-dom` stays really lean and renders whatever you ask it to render.

Here is a requirebin to demonstrate the approach: http://requirebin.com/?gist=29e810cf36bea1ec44cb

**Comment by neonstalwart** [comment]
@ih2502mk yes, this is closer to how we think it should be done.  ideally you don't want to have your application logic need to manage the fake item or understand how long the transition takes, etc.  instead, you may want to generalize this pattern and make it part of the responsibility of your render code by using a Widget as the container for the list.  the widget can be the one to manage the fake item.  it can detect when an item is removed and insert a fake item and can remove it when the transition is over.
